File: modules/dms-vision/ios/DmsVisionSupport.swift
// Shared helpers for the DmsVision module: bundled-resource resolution, the landmark
// buffer-frame -> upright-frame transform, camera intrinsics and the thermal probe.
//
// Conventions (must match dms/gaze_inputs.js and the research pipeline):
//   * landmarks are MediaPipe normalized (x / W, y / H, z / W) of the UPRIGHT image;
//   * the serialized landmark buffer is 478 * 3 little-endian float32, point-major;
//   * focalScale = fx / uprightWidth, principal point assumed at the frame centre.

import ExpoModulesCore
import Foundation
import AVFoundation
import CoreMedia
import UIKit
import simd

let kDmsNumLandmarks = 478

// MARK: - Errors

internal final class DmsVisionException: GenericException<String> {
  override var reason: String { return param }
}

// MARK: - Bundled resources

/// Resolves a file that ships inside the module's `DmsVision` resource bundle.
///
/// CocoaPods puts a `resource_bundles` bundle inside the pod's framework when the app is
/// built with frameworks and inside the app bundle when it is built with static libraries,
/// so both locations are searched, then the plain main bundle as a last resort.
internal enum DmsVisionBundle {
  private static let bundleName = "DmsVision"

  static func path(forResource name: String, ofType ext: String) -> String? {
    var candidates: [Bundle] = []
    let ownBundle = Bundle(for: DmsVisionBundleToken.self)
    for host in [ownBundle, Bundle.main] {
      if let url = host.url(forResource: bundleName, withExtension: "bundle"),
         let bundle = Bundle(url: url) {
        candidates.append(bundle)
      }
    }
    candidates.append(ownBundle)
    candidates.append(Bundle.main)

    for bundle in candidates {
      if let path = bundle.path(forResource: name, ofType: ext),
         FileManager.default.fileExists(atPath: path) {
        return path
      }
    }
    return nil
  }

  static func require(_ name: String, _ ext: String) throws -> String {
    guard let path = path(forResource: name, ofType: ext) else {
      throw DmsVisionException("Bundled resource \(name).\(ext) not found; check the DmsVision podspec resource_bundles")
    }
    return path
  }
}

/// Only exists so `Bundle(for:)` can find the framework this file was compiled into.
private final class DmsVisionBundleToken {}

// MARK: - Orientation

/// Clockwise rotation, in degrees, that must be applied to the delivered camera buffer for
/// its contents to be upright.
///
/// iPhone camera buffers are delivered in the sensor's native landscape orientation for both
/// the front and the back camera (AVFoundation's `videoRotationAngle` maps
/// `.landscapeRight` -> 0 and `.portrait` -> 90 for every device), so a portrait device needs
/// 90 degrees clockwise. Mirroring is disabled explicitly, so no mirrored variant applies.
internal func dmsRotationDegrees(for orientation: UIDeviceOrientation) -> Int {
  switch orientation {
  case .landscapeLeft: return 0
  case .landscapeRight: return 180
  case .portraitUpsideDown: return 270
  default: return 90       // .portrait, .faceUp, .faceDown, .unknown
  }
}

internal func dmsOrientationName(for orientation: UIDeviceOrientation) -> String {
  switch orientation {
  case .landscapeLeft: return "landscapeLeft"
  case .landscapeRight: return "landscapeRight"
  case .portraitUpsideDown: return "portraitUpsideDown"
  default: return "portrait"
  }
}

/// `UIImage.Orientation` whose MediaPipe rotation equals `degrees` clockwise.
///
/// Verified against mediapipe/tasks/ios/vision/core/sources/MPPVisionTaskRunner.mm, which maps
/// `.right` -> a 270 degree counter-clockwise NormalizedRect rotation (= 90 clockwise),
/// `.down` -> 180 and `.left` -> 90 counter-clockwise (= 270 clockwise).
internal func dmsImageOrientation(forClockwiseDegrees degrees: Int) -> UIImage.Orientation {
  switch ((degrees % 360) + 360) % 360 {
  case 90: return .right
  case 180: return .down
  case 270: return .left
  default: return .up
  }
}

// MARK: - Landmark transform

/// Rotates MediaPipe's normalized landmarks from the UNROTATED buffer frame into the upright
/// frame and serializes them as little-endian float32.
///
/// MediaPipe Tasks returns landmarks in the unrotated input frame even when a rotation is
/// requested. Measured with mediapipe 0.10.35 and this exact `face_landmarker.task`
/// (scratchpad probe, 2026-09-18): applying the transform below reproduces the landmarks of a
/// physically pre-rotated image to 0.003-0.005 normalized units (detector noise), while the
/// raw buffer-frame points differ by 0.28-0.45.
///
/// `z` is NOT rescaled. MediaPipe scales `z` by the length of the projected face-ROI x-axis
/// measured in the output (x / W, y / H) frame (landmark_projection_calculator.cc,
/// `CalculateZScale`), and the ROI follows the face, so for a 90/270 degree rotation `z` is
/// already divided by the upright width. The same probe measured max |z_rotated - z_reference|
/// = 0.0004-0.03 against 0.04-0.08 for the "z * bufferWidth / bufferHeight" alternative.
///
/// `interleaved` is (x, y, z) * 478 in the unrotated buffer frame.
internal func dmsSerializeLandmarks(_ interleaved: [Float], rotationDegrees: Int) -> Data {
  var out = interleaved
  let rotation = ((rotationDegrees % 360) + 360) % 360
  if rotation != 0 {
    var i = 0
    while i + 2 < out.count {
      let bx = out[i], by = out[i + 1]
      switch rotation {
      case 90:  out[i] = 1.0 - by; out[i + 1] = bx
      case 180: out[i] = 1.0 - bx; out[i + 1] = 1.0 - by
      case 270: out[i] = by;       out[i + 1] = 1.0 - bx
      default:  break
      }
      i += 3
    }
  }
  return out.withUnsafeBufferPointer { Data(buffer: $0) }
}

// MARK: - Intrinsics

internal struct DmsIntrinsics {
  /// Focal length in pixels of the delivered (unrotated) buffer.
  var fx: Double = 0
  var fy: Double = 0
  /// Principal point in pixels of the delivered (unrotated) buffer.
  var cx: Double = 0
  var cy: Double = 0
  var bufferWidth: Int = 0
  var bufferHeight: Int = 0
  /// "intrinsics" (sample-buffer attachment), "fov" (videoFieldOfView) or "default".
  var source: String = "default"

  static let defaultHorizontalFovDegrees: Double = 70.0

  static func fromFieldOfView(_ hfovDegrees: Double, width: Int, height: Int, source: String) -> DmsIntrinsics {
    var out = DmsIntrinsics()
    let hfov = (hfovDegrees > 1.0 && hfovDegrees < 179.0) ? hfovDegrees : defaultHorizontalFovDegrees
    let fx = Double(width) / (2.0 * tan(hfov * Double.pi / 360.0))
    out.fx = fx
    out.fy = fx                       // square pixels
    out.cx = Double(width) / 2.0
    out.cy = Double(height) / 2.0
    out.bufferWidth = width
    out.bufferHeight = height
    out.source = (hfovDegrees > 1.0 && hfovDegrees < 179.0) ? source : "default"
    return out
  }

  /// Upright image size after `rotationDegrees` clockwise.
  func uprightSize(rotationDegrees: Int) -> (width: Int, height: Int) {
    let swapped = (((rotationDegrees % 360) + 360) % 360) % 180 != 0
    return swapped ? (bufferHeight, bufferWidth) : (bufferWidth, bufferHeight)
  }

  /// `fx / uprightWidth`. A 90/270 degree rotation swaps the image axes, so the upright
  /// horizontal focal length is the buffer's vertical one.
  func focalScale(rotationDegrees: Int) -> Double {
    let swapped = (((rotationDegrees % 360) + 360) % 360) % 180 != 0
    let fxUpright = swapped ? fy : fx
    let uprightWidth = Double(swapped ? bufferHeight : bufferWidth)
    guard uprightWidth > 0, fxUpright.isFinite, fxUpright > 0 else { return 0 }
    return fxUpright / uprightWidth
  }
}

/// Reads `kCMSampleBufferAttachmentKey_CameraIntrinsicMatrix` if the connection delivers it.
internal func dmsIntrinsicsFromSampleBuffer(_ sampleBuffer: CMSampleBuffer, width: Int, height: Int) -> DmsIntrinsics? {
  guard let raw = CMGetAttachment(sampleBuffer,
                                  key: kCMSampleBufferAttachmentKey_CameraIntrinsicMatrix,
                                  attachmentModeOut: nil) as? Data,
        raw.count >= MemoryLayout<matrix_float3x3>.size else {
    return nil
  }
  var matrix = matrix_float3x3()
  withUnsafeMutableBytes(of: &matrix) { destination in
    raw.copyBytes(to: destination.bindMemory(to: UInt8.self),
                  from: 0..<MemoryLayout<matrix_float3x3>.size)
  }
  let fx = Double(matrix.columns.0.x)
  let fy = Double(matrix.columns.1.y)
  guard fx.isFinite, fy.isFinite, fx > 1.0, fy > 1.0 else { return nil }
  var out = DmsIntrinsics()
  out.fx = fx
  out.fy = fy
  out.cx = Double(matrix.columns.2.x)
  out.cy = Double(matrix.columns.2.y)
  out.bufferWidth = width
  out.bufferHeight = height
  out.source = "intrinsics"
  return out
}

// MARK: - Thermal

internal func dmsThermalStateName() -> String {
  switch ProcessInfo.processInfo.thermalState {
  case .nominal: return "nominal"
  case .fair: return "fair"
  case .serious: return "serious"
  case .critical: return "critical"
  @unknown default: return "unknown"
  }
}

File: modules/dms-vision/android/src/main/java/expo/modules/dmsvision/DmsVisionSupport.kt
package expo.modules.dmsvision

import android.content.Context
import android.graphics.Rect
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.os.Build
import android.os.PowerManager
import android.util.Size
import android.util.SizeF
import android.view.Surface
import expo.modules.kotlin.exception.CodedException
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

const val DMS_NUM_LANDMARKS = 478

class DmsVisionException(message: String) : CodedException(message)

// ---------------------------------------------------------------------------------------------
// Bundled assets
// ---------------------------------------------------------------------------------------------

/**
 * Reads a file from the module's own `src/main/assets` (merged into the app's assets at build
 * time). Nothing is downloaded and nothing is copied to the cache: MediaPipe takes the model as a
 * direct [ByteBuffer] and ONNX Runtime takes it as a byte array, so asset compression is
 * irrelevant.
 */
object DmsVisionAssets {
  fun readBytes(context: Context, name: String): ByteArray {
    try {
      context.assets.open(name).use { input ->
        val out = ByteArrayOutputStream(maxOf(input.available(), 1 shl 16))
        input.copyTo(out)
        return out.toByteArray()
      }
    } catch (e: Exception) {
      throw DmsVisionException("bundled asset $name is missing: ${e.message}")
    }
  }

  fun readDirectBuffer(context: Context, name: String): ByteBuffer {
    val bytes = readBytes(context, name)
    val buffer = ByteBuffer.allocateDirect(bytes.size).order(ByteOrder.nativeOrder())
    buffer.put(bytes)
    buffer.rewind()
    return buffer
  }
}

// ---------------------------------------------------------------------------------------------
// Orientation
// ---------------------------------------------------------------------------------------------

/**
 * Maps the physical device orientation reported by `OrientationEventListener` to the
 * `Surface.ROTATION_*` value that `ImageAnalysis.setTargetRotation` expects. Verbatim from the
 * CameraX "Rotations" guidance.
 */
fun dmsSurfaceRotation(orientationDegrees: Int): Int = when {
  orientationDegrees >= 45 && orientationDegrees < 135 -> Surface.ROTATION_270
  orientationDegrees >= 135 && orientationDegrees < 225 -> Surface.ROTATION_180
  orientationDegrees >= 225 && orientationDegrees < 315 -> Surface.ROTATION_90
  else -> Surface.ROTATION_0
}

/** Names match the iOS side so the JS contract is platform-independent. */
fun dmsOrientationName(surfaceRotation: Int): String = when (surfaceRotation) {
  Surface.ROTATION_90 -> "landscapeLeft"
  Surface.ROTATION_180 -> "portraitUpsideDown"
  Surface.ROTATION_270 -> "landscapeRight"
  else -> "portrait"
}

// ---------------------------------------------------------------------------------------------
// Landmark transform
// ---------------------------------------------------------------------------------------------

/**
 * Rotates MediaPipe's normalized landmarks from the UNROTATED buffer frame into the upright frame
 * and serializes them as little-endian float32, point-major (x, y, z) * 478.
 *
 * MediaPipe Tasks returns landmarks in the unrotated input frame even when
 * `ImageProcessingOptions.rotationDegrees` is set. Measured with mediapipe 0.10.35 and this exact
 * `face_landmarker.task` (2026-09-18): the transform below reproduces the landmarks of a
 * physically pre-rotated image to 0.003-0.005 normalized units, while the raw buffer-frame points
 * differ by 0.28-0.45.
 *
 * `z` is NOT rescaled - MediaPipe already scales it by the projected face-ROI x-axis measured in
 * the output (x / W, y / H) frame, which for a 90/270 degree rotation equals the upright width.
 * See docs/dms/NATIVE_LAYER.md.
 */
fun dmsSerializeLandmarks(interleaved: FloatArray, rotationDegrees: Int): ByteArray {
  val rotation = ((rotationDegrees % 360) + 360) % 360
  val out = ByteBuffer.allocate(interleaved.size * 4).order(ByteOrder.LITTLE_ENDIAN)
  var i = 0
  while (i + 2 < interleaved.size) {
    val bx = interleaved[i]
    val by = interleaved[i + 1]
    val ux: Float
    val uy: Float
    when (rotation) {
      90 -> { ux = 1.0f - by; uy = bx }
      180 -> { ux = 1.0f - bx; uy = 1.0f - by }
      270 -> { ux = by; uy = 1.0f - bx }
      else -> { ux = bx; uy = by }
    }
    out.putFloat(ux)
    out.putFloat(uy)
    out.putFloat(interleaved[i + 2])
    i += 3
  }
  return out.array()
}

// ---------------------------------------------------------------------------------------------
// Camera intrinsics
// ---------------------------------------------------------------------------------------------

/**
 * Focal length and principal point in pixels of the DELIVERED (unrotated) analysis buffer.
 * `source` is "intrinsics" when it was derived from the camera characteristics and "default" when
 * the 70 degree fallback horizontal field of view was used.
 */
data class DmsIntrinsics(
  val fx: Double,
  val fy: Double,
  val cx: Double,
  val cy: Double,
  val bufferWidth: Int,
  val bufferHeight: Int,
  val source: String
) {
  fun uprightSize(rotationDegrees: Int): Pair<Int, Int> {
    val swapped = (((rotationDegrees % 360) + 360) % 360) % 180 != 0
    return if (swapped) Pair(bufferHeight, bufferWidth) else Pair(bufferWidth, bufferHeight)
  }

  /** `fx / uprightWidth`; a 90/270 degree rotation swaps the image axes. */
  fun focalScale(rotationDegrees: Int): Double {
    val swapped = (((rotationDegrees % 360) + 360) % 360) % 180 != 0
    val fxUpright = if (swapped) fy else fx
    val uprightWidth = (if (swapped) bufferHeight else bufferWidth).toDouble()
    if (uprightWidth <= 0.0 || !fxUpright.isFinite() || fxUpright <= 0.0) return 0.0
    return fxUpright / uprightWidth
  }

  companion object {
    const val DEFAULT_HORIZONTAL_FOV_DEGREES = 70.0

    fun fromFieldOfView(hfovDegrees: Double, width: Int, height: Int, source: String): DmsIntrinsics {
      val usable = hfovDegrees > 1.0 && hfovDegrees < 179.0
      val hfov = if (usable) hfovDegrees else DEFAULT_HORIZONTAL_FOV_DEGREES
      val fx = width.toDouble() / (2.0 * Math.tan(Math.toRadians(hfov / 2.0)))
      return DmsIntrinsics(fx, fx, width / 2.0, height / 2.0, width, height,
        if (usable) source else "default")
    }

    fun default(width: Int, height: Int): DmsIntrinsics =
      fromFieldOfView(DEFAULT_HORIZONTAL_FOV_DEGREES, width, height, "default")
      .copy(source = "default")
  }
}

/**
 * Computes the front camera's intrinsics for a delivered buffer of [width] x [height] from the
 * Camera2 characteristics, following the AOSP centre-crop rule:
 *
 *   mmPerPx   = physicalSize / pixelArraySize
 *   active_mm = activeArray * mmPerPx
 *   the stream is a CENTRED crop of the active array with the stream's aspect ratio
 *   fx_px     = focalLength_mm * outW / cropW_mm
 *
 * `LENS_INTRINSIC_CALIBRATION` is deliberately NOT used: AOSP marks it optional, it is commonly
 * null or zeroed on front cameras, and its units are pixels in the pre-correction active array
 * coordinate system, which is a different frame from the delivered buffer.
 *
 * The characteristics are read through `CameraManager` rather than CameraX's
 * `Camera2CameraInfo`, which is an opt-in experimental interop API; the first front-facing camera
 * id is the one `CameraSelector.DEFAULT_FRONT_CAMERA` resolves to.
 */
fun dmsFrontCameraIntrinsics(context: Context, width: Int, height: Int): DmsIntrinsics {
  if (width <= 0 || height <= 0) return DmsIntrinsics.default(maxOf(width, 1), maxOf(height, 1))
  try {
    val manager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
    for (id in manager.cameraIdList) {
      val characteristics = manager.getCameraCharacteristics(id)
      if (characteristics.get(CameraCharacteristics.LENS_FACING) !=
        CameraCharacteristics.LENS_FACING_FRONT
      ) {
        continue
      }
      val focalLengths = characteristics.get(CameraCharacteristics.LENS_INFO_AVAILABLE_FOCAL_LENGTHS)
      val physical: SizeF? = characteristics.get(CameraCharacteristics.SENSOR_INFO_PHYSICAL_SIZE)
      val pixelArray: Size? = characteristics.get(CameraCharacteristics.SENSOR_INFO_PIXEL_ARRAY_SIZE)
      val activeArray: Rect? = characteristics.get(CameraCharacteristics.SENSOR_INFO_ACTIVE_ARRAY_SIZE)
      if (focalLengths == null || focalLengths.isEmpty() || physical == null ||
        pixelArray == null || activeArray == null ||
        pixelArray.width <= 0 || pixelArray.height <= 0 ||
        activeArray.width() <= 0 || activeArray.height() <= 0 ||
        physical.width <= 0f || physical.height <= 0f
      ) {
        return DmsIntrinsics.default(width, height)
      }
      val focal = focalLengths[0].toDouble()
      if (focal <= 0.0) return DmsIntrinsics.default(width, height)

      val mmPerPxX = physical.width.toDouble() / pixelArray.width.toDouble()
      val mmPerPxY = physical.height.toDouble() / pixelArray.height.toDouble()
      val activeWidthMm = activeArray.width() * mmPerPxX
      val activeHeightMm = activeArray.height() * mmPerPxY

      val outAspect = width.toDouble() / height.toDouble()
      val activeAspect = activeWidthMm / activeHeightMm
      val cropWidthMm: Double
      val cropHeightMm: Double
      if (outAspect > activeAspect) {
        cropWidthMm = activeWidthMm
        cropHeightMm = activeWidthMm / outAspect
      } else {
        cropHeightMm = activeHeightMm
        cropWidthMm = activeHeightMm * outAspect
      }
      if (cropWidthMm <= 0.0 || cropHeightMm <= 0.0) return DmsIntrinsics.default(width, height)

      val fx = focal * width.toDouble() / cropWidthMm
      val fy = focal * height.toDouble() / cropHeightMm
      if (!fx.isFinite() || !fy.isFinite() || fx <= 1.0 || fy <= 1.0) {
        return DmsIntrinsics.default(width, height)
      }
      return DmsIntrinsics(fx, fy, width / 2.0, height / 2.0, width, height, "intrinsics")
    }
  } catch (_: Exception) {
    // fall through to the default
  }
  return DmsIntrinsics.default(width, height)
}

// ---------------------------------------------------------------------------------------------
// Thermal
// ---------------------------------------------------------------------------------------------

/** "nominal" | "fair" | "serious" | "critical" | "unknown" (below API 29). */
fun dmsThermalStateName(context: Context): String {
  if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return "unknown"
  val power = context.getSystemService(Context.POWER_SERVICE) as? PowerManager ?: return "unknown"
  return dmsThermalStatusName(power.currentThermalStatus)
}

fun dmsThermalStatusName(status: Int): String = when (status) {
  PowerManager.THERMAL_STATUS_NONE -> "nominal"
  PowerManager.THERMAL_STATUS_LIGHT, PowerManager.THERMAL_STATUS_MODERATE -> "fair"
  PowerManager.THERMAL_STATUS_SEVERE -> "serious"
  PowerManager.THERMAL_STATUS_CRITICAL,
  PowerManager.THERMAL_STATUS_EMERGENCY,
  PowerManager.THERMAL_STATUS_SHUTDOWN -> "critical"
  else -> "unknown"
}

fun dmsLowPowerMode(context: Context): Boolean {
  val power = context.getSystemService(Context.POWER_SERVICE) as? PowerManager ?: return false
  return try {
    power.isPowerSaveMode
  } catch (_: Exception) {
    false
  }
}
